### ISSUE TITLE
StaticType Rule Bugs

### ISSUE THREAD

**Issue by MeirionHughes** [MeirionHughes/aurelia-template-lint#35]
Please report any issues you find and I will try to collate and fix them as I can. 

* [x] ~~Column offsets are always 0~~  
* [x] ~~Order of Attributes in same element can break bindings to a locals~~ 
* [x] ~~Ancestor access doesn't check types i.e. `$parent`~~
* [x] ~~Capitalised filename should resolve (limited to Foo*ViewModel* currently)~~
* [x] ~~Doesn't follow camel-case convention file <-> class~~
* [x] ~~Doesn't support Mapping ex:  `repeat.for="[id, customer] of customers"`~~
* [x] ~~Complains it cannot find converter as field member~~
* [x] ~~Should add `$index`, `$odd` `$even` inside repeat.for as known variables~~
* [x] ~~context variables aren't overridden in nested controllers (same local name)~~
* [x] ~~Doesn't recognise interfaces.~~ 
* [x] ~~repeat.for doesn't work with interfaces imported from other files~~
* [x] ~~Template is not checked if the view model class name is different~~
* [x] ~~Traverses into private fields (should be warning).~~ 
* [x] ~~Doesn't support interpolation with attribute value~~
* [x] ~~Doesn't support Keyed Access~~
* [x] ~~Doesn't support negation operator on if.bind~~ 
* [x] ~~Bindable attributes passed to custom elements should be checked.~~
* [x] ~~Custom elements should be checked.~~

* ~~_Doesn't support Generics and use of Generic member types_~~ [moved to feature](https://github.com/MeirionHughes/aurelia-template-lint/issues/50)
* ~~_Doesn't support Value Converters (different Type output)_~~ [moved to feature](https://github.com/MeirionHughes/aurelia-template-lint/issues/51)

**Comment by zakjan** [comment]
I tried to activate it with `useStaticTyping = true` but gulp now just hangs up for me. No other tasks are started. Version 0.7.2

```
$ gulp watch
[15:22:06] Using gulpfile xxx/gulpfile.js
[15:22:06] Starting 'build'...
[15:22:06] Starting 'lint-html'...
```

**Comment by zakjan** [comment]
repeat.for doesn't work with interfaces imported from other files

item.ts
```
export interface Item {}
```

page.ts
```
import {Item} from "./item";

export class PageViewModel {
    items: Item[];
}
```

page.html
```
<template>
    <template repeat.for="item of items">
        ${item}
    </template>
</template>
```

lint output:
```
[01:07:19] WARNING: cannot find 'item' in type 'PageViewModel' Ln 3 Col 0 /src/pages/page.html
```

**Comment by zakjan** [comment]
Template is not checked if the view model class name is different from the view and view model file names.

src/profiles/list.ts
```
export class ProfilesListViewModel {}
```

src/profiles/list.html
```
<template>
    ${item}
</template>
```

There is no linter issue found.

Also about your point "Doesn't follow camel-case convention file <-> class", I think case convention shouldn't be enforced, the same basename of both files should be enough to check them.

**Comment by zakjan** [comment]
No custom elements are checked.

item.ts
```
import {bindable} from "aurelia-templating";

export class ItemCustomElement {
    @bindable item: string;
}
```

item.html
```
<template>
    ${itemm}
</template>
```

There is no linter issue found.

**Comment by zakjan** [comment]
Bindable attributes passed to custom elements should be checked.

item.ts
```
import {bindable} from "aurelia-templating";

export class ItemCustomElement {
    @bindable value: string;
}
```

item.html
```
<template>
    ${item}
</template>
```

page.ts
```
export class PageViewModel {
    value: number;
}
```

page.html
```
<template>
    <require from="./item"></require>

    <item value.bind="value"></item>
</template>
```

**Comment by zakjan** [comment]
An idea to think about: How about to optionally allow only strings (and maybe numbers) in string interpolations?

```
export class PageViewModel {
    item: {name: string};
}

<template>
    <!-- fail -->
    ${item}
    <div title="Item: ${item}"></div>
    <!-- ok -->
    ${item.name}
    <div title="Item: ${item.name}"></div>
</template>
```

It could prevent unintentional implicit casting to strings and displaying strings such as `[object Object]`.

**Comment by atsu85** [comment]
I'd certainly use that.

**Comment by MeirionHughes** [comment]
>  ...allow only strings (and maybe numbers) in string interpolations?
yeah that's good idea; could you lift that out as a feature request? I don't think its a bug _per se_.

**Comment by zakjan** [comment]
Yeah, of course. #44

**Comment by zakjan** [comment]
```
export class PageViewModel {
    value: string;
}

<template>
    <template if.bind="!value">
        ...
    </template>
</template>
```

linter output:

```
[20:48:18] WARNING: cannot find 'undefined' in type 'PageViewModel' Ln 2 Col 0 /src/pages/profiles/create/page.html
```

**Comment by MeirionHughes** [comment]
with regard to attribute ordering and using a local before it has been declared; 
the issue is this: 
```html
<option model.bind = "employee.id" repeat.for="employee of employees"  > 
```

which can be fixed by using a more (in my view) logical ordering: 
```html
<option repeat.for="employee of employees"  model.bind = "employee.id" > 
```

given there are so many issues to work though, I won't be fixing this as yet.

**Comment by atsu85** [comment]
> which can be fixed by using a more (in my view) logical ordering: 

@EisenbergEffect, do I remember correctly, that different browsers gibe attributes in different order? If I remember correctly, then one browser gave it in the order it was written in source code, and some other browser gave it alphabetically? If so, then we shouldn't rely on the order of attributes, but rather set them on different elements?

**Comment by EisenbergEffect** [comment]
Yes. I don't think you can rely on attribute order. That's one reason why you can't have two template controllers on the same element and have to nest a template.

**Comment by EisenbergEffect** [comment]
But, the above example should work because the template controller overrides everything else and causes the scope to be created.

**Comment by atsu85** [comment]
@EisenbergEffect, I don't understand Your last message. Do You mean, that the order of `repeat.for` and `model.bind` shouldn't matter, as 
> template controller overrides everything else and causes the scope to be created.

(that would contradict what @MeirionHughes [mentioned earlier](https://github.com/MeirionHughes/aurelia-template-lint/issues/35#issuecomment-229642049))
or do You mean, that in case `repeat.for` comes before `model.bind` it should work (as @MeirionHughes said)?

**Comment by EisenbergEffect** [comment]
The order should not matter. The presence of a template controller overrides everything else on the element. However, it makes more sense to have the template controller first.

**Comment by MeirionHughes** [comment]
Resolved; realized I could simply reorder them now that I pre-process into an AST.

**Comment by atsu85** [comment]
So this is the first feature, that modifies the output compared to the input?

**Comment by MeirionHughes** [comment]
Nope, nothing changes. The analysis code is basically one bit state machine; I've simply re-ordered the attributes so the template controllers are examined first; presumably only the controllers add local variables. This means if you have `<option model.bind = "employee.id" repeat.for="employee of employees"  >` in your code; it will correctly handle `model.bind` and know about `employee`

**Comment by niieani** [comment]
Linting the navigation-skeleton with StaticType analysis shows the error:

```
WARNING in ./src/users.html
[9, 11]: cannot find 'image' in type 'Users'

 @ ./src/users.ts 95:0-31
 @ ./src/app.ts
 @ ./src/main.ts
 @ multi app
```

The source of this is `blur-image.bind="image"`:

```html
<canvas class="header-bg" width="250" height="70" blur-image.bind="image"></canvas>
<div class="avatar">
  <img src.bind="user.avatar_url" crossorigin ref="image"/>
</div>
```

The thing is, it's binding to the `ref` on the `img` below. Also, the whole thing is in a `repeat.for="user of users"` loop, so `image` property belongs on the `IUser` interface, not on the `Users` view-model.

**Comment by RomkeVdMeulen** [comment]
I'm getting false positives on vars added through `ref` like @niieani mentioned. Also a lot of false positives on things added through aurelia-validation's `validation-errors` attribute.

```html
<form id="new" validation-errors="errors.bind: newErrors; controller.bind: newValidationController">
	<input type="text" value.bind="new.name & validate"/>
	<input type="submit"/>

	<p repeat.for="newError of newErrors" class="warning">
		${newError.error.message}
	</p>
</form>
```

gives:

```
[11:07:15] WARNING: cannot find 'newErrors' in type 'MyEditor' Ln 5
```

The simple binding `validation-errors.bind="errors"` also gives these false positives.
